### ISSUE TITLE
[Interface cleanup, Discussion] DeterministicError marker trait

### ISSUE THREAD

**Issue by nearmax** [nearprotocol/nearcore#3326]
# Motivation

Our current runtime error hierarchy is extremely complex, unfortunately we have to differentiate several kinds of errors:
* Deterministic and non-deterministic errors. Deterministic errors are caused during transaction execution and in most cases result of user executing an invalid transaction. 
* Transient and persistent non-deterministic errors. Transient errors go away when the operation is retried (e.g. disk read), but non-deterministic errors do not go away.

Currently we treat all transient errors as persistent errors, i.e. we do not retry any operations, like disk access in the runtime.

Also, depending on the context an error can be viewed as deterministic or non-deterministic. E.g. If we encounter missing state trie node during regular contract execution it is a sign of corrupted disk and is a non-deterministic persistent error that causes node process to be terminated. However, if we encounter missing state trie node during challenge validation it is a sign of corrupted challenge and is treated as deterministic error that causes challenge to be rejected.

Historically, we had bugs when non-deterministic error was processed as deterministic (the node would consider it a transaction failure instead of terminating the node process). The most recent example is here: https://github.com/nearprotocol/nearcore/pull/3257/

# Proposal

TLDR: Use the same "poisoning mechanism" that `!Sync` uses to propagate thread unsafety through the type hierarchy.

We can leverage Rust type system to minimize the risk of miscategorizing deterministic errors as non-determinstic. For example if error type `A` is declared to be non-deterministic then it should not be possible to declare type` enum B { A(A), C(C), D(D), ..}` as deterministic. It can be done similarly to how Rust uses `Sync` and `Send` marker traits. For all deterministic errors types that are deterministic we would add decorator `#[derive(DeterministicError)]` which would implement `DeterministicError` marker trait. 

For example the following code should fail to compile:
```rust
#[derive(DeterministicError)]
enum ContractError {
  OutOfBalance(OutOfBalance),
  StorageError(StorageError),
  ...
}

#[derive(DeterministicError)]
enum StorageError {
  MissingAccount(AccountId),
  MissingNode(MissingNodeError)
}

struct MissingNodeError {
...
}

impl !DeterminsticError for MissingNodeError {}
```

**Comment by frol** [comment]
FYI, negative bounds are not available in Rust (those `!Sync` and `!Send` are special cases)

**Comment by nearmax** [comment]
Closing then.
